libraries/rib-base/src/main/java/com/badoo/ribs/clienthelper/childawareness/ChildAwareCallbackInfo.kt

```diff
@@ -0,0 +1,97 @@
+package com.badoo.ribs.clienthelper.childawareness
+
+import androidx.lifecycle.Lifecycle
+import com.badoo.ribs.core.Node
+import com.badoo.ribs.core.Rib
+import com.badoo.ribs.core.lifecycle.MinimumCombinedLifecycle
+import kotlin.reflect.KClass
+import kotlin.reflect.cast
+import kotlin.reflect.safeCast
+
+internal sealed class ChildAwareCallbackInfo {
+    abstract val mode: Mode
+
+    class Single<T : Rib>(
+        private val child: KClass<T>,
+        private val callback: ChildCallback<T>,
+        private val parentLifecycle: Lifecycle,
+        override val mode: Mode,
+    ) : ChildAwareCallbackInfo() {
+
+        fun invokeIfRequired(newNode: Node<*>) {
+            val castedNode = child.safeCast(newNode)
+            if (castedNode != null) {
+                @Suppress("UNCHECKED_CAST")
+                callback(
+                    MinimumCombinedLifecycle(parentLifecycle, newNode.lifecycle).lifecycle,
+                    castedNode
+                )
+            }
+        }
+
+        fun invokeIfRequired(activeNodes: List<Node<*>>) {
+            activeNodes.forEach { node ->
+                invokeIfRequired(node)
+            }
+        }
+
+    }
+
+    @Suppress("LongParameterList")
+    class Double<T1 : Rib, T2 : Rib>(
+        private val child1: KClass<T1>,
+        private val child2: KClass<T2>,
+        private val callback: ChildrenCallback<T1, T2>,
+        private val parentLifecycle: Lifecycle,
+        override val mode: Mode,
+    ) : ChildAwareCallbackInfo() {
+
+        init {
+            require(child1 != child2) { "Registered classes should be different ($child1, $child2)" }
```

Comment:
I can imagine a parent having the same items (same child class) in a container. I wonder if it's a hard limit that we can't solve, or just seemed logical at first to exclude it?

Reply:
Just seemed logical at first to exclude it, will remove it.